react/components/Slider.js

```diff
@@ -1,7 +1,393 @@
+  }
+
+  static events = ['onMouseUp', 'onMouseDown', 'onMouseLeave', 'onMouseMove']
+
+  constructor(props) {
+    super(props)
+
+    this.drag = {
+      startX: 0,
+      endX: 0,
+      startY: 0,
+      letItGo: null
+    }
+
+    this.transformProperty = this.getTransformProperty()
+  }
+
+  getTransformProperty = () => {
+    const transform = typeof window !== 'undefined' && window.document.documentElement.style.transform
+
+    if (typeof transform === 'string') {
+      return 'transform'
+    }
+    return 'WebkitTransform'
+  }
+
+  componentDidMount() {
+    this.init()
+    this.onResize = debounce(this.handleResize, this.props.resizeDebounce)
+    window.addEventListener('resize', this.onResize)
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('resize', this.onResize)
+  }
+
+  componentDidUpdate() {
+    this.init()
+  }
+
+  getStylingTransition = (easing, duration = 0) => ({
+    webkitTransition: `all ${duration}ms ${easing}`,
+    transition: `all ${duration}ms ${easing}`
+  })
+
+  enableTransition = (extraStyles = {}) => {
+    const { easing, duration } = this.props
+
+    this.setStyle(this.sliderFrame, {
+      ...this.getStylingTransition(easing, duration),
+      ...extraStyles
+    })
+  }
+
+  disableTransition = (extraStyles = {}) => {
+    const { easing } = this.props
+    this.setStyle(this.sliderFrame, {
+      ...this.getStylingTransition(easing),
+      ...extraStyles
+    })
+  }
+
+  init = () => {
+    const { draggable, currentSlide, loop } = this.props
+    this.setSelectorWidth()
+    this.setInnerElements()
+    this.perPage = resolveSlidesNumber(this.props.perPage)
+    this.enableTransition({
+      width: `${(this.selectorWidth / this.perPage) * this.innerElements.length}px`,
+      ...(draggable ? { cursor: '-webkit-grab' } : {}),
+    })
+
+    this.innerElements.forEach(el => {
+      this.setStyle(el, {
+        width: `${100 / this.innerElements.length}%`
+      })
+    })
+    let newCurrentSlide = loop ? currentSlide % this.totalSlides : Math.min(Math.max(currentSlide, 0), this.totalSlides - this.perPage)
+    this.slideToCurrent(false, newCurrentSlide)
+  }
+
+  handleResize = () => {
+    const { perPage, currentSlide, onChangeCurrentSlide } = this.props
+    this.perPage = resolveSlidesNumber(perPage)
+    const newCurrentSlide = Math.floor(currentSlide / this.perPage) * this.perPage
+
+    this.selectorWidth = this.selector.getBoundingClientRect().width
+    this.setStyle(this.sliderFrame, {
+      width: `${(this.selectorWidth / this.perPage) * this.innerElements.length}px`
+    })
+    
+    if (currentSlide !== newCurrentSlide) {
+      onChangeCurrentSlide(newCurrentSlide)
+    }
+
+    this.slideToCurrent(false, newCurrentSlide)
+    this.forceUpdate()
+  }
+
+  setSelectorWidth = () => {
+    this.selectorWidth = this.selector.getBoundingClientRect().width
+  }
+
+  setInnerElements = () => {
+    this.innerElements = [].slice.call(this.sliderFrame.children)
```

Comment:
Could use the `Array` object directly, instead of an instance.

Reply:
> Also, one other thing I noticed is that you require that the user passes in how many slides they want on each page, which is totally valid if the user _knows_ how many slides they want.
> 
> But there are cases where we _don't know_ how many slides there will be, and want to fit as much as we can on a single page, have you already thought of that?

If you're talking about `howManySlides` prop I explained what happened already, but if it is about `perPage` you have to pass how many slides should fit on the screen, there you can pass a number or you can pass an object with breakpoints. These values will be ignored in the first render, since it won't know the size of the screen on SSR. But anyway... I have to write some docs and have better names to theses props

Reply:
I don't know why github replied here 😕

Reply:
> Could use the `Array` object directly, instead of an instance.

I can't, because `this.sliderFrame.children` is not an array, is a `HTMLCollection`.

Reply:
I was talking about doing something like `Array.prototype.slice.call(...)`